4 feature edit maintenance event detail

aaronthebald opened the pull request:
# What it Does
* Closes #4
* Using @Binding you can now swipe on an event, and press the edit button. This hands the selectedEvent over to the EditMaintenanceEventDetailsView. From there the text fields are prepopulated with the data from the selectedEvent. Once updated, using the @DocumentID to find the correct document in firebase the function overwrites the old data with the new. Then the function calls getMaintenanceEvents() to update the array of current events.

# How I Tested
* swipe on existing event.
* press edit
* update text fields in the .sheet.
* Press update.
etc...

# Notes
* Anything else that should be noted about how you implemented this feature?

# Screenshot
* Add a screenshot of your new feature! OR show a screen recording of it in action. (On the simulator press Cmd + R to record, and the stop button when done). **This video should be no longer than 30 seconds.**

aaronthebald committed:
Added features that allow the user to update Maintenance event

aaronthebald committed:
Attempting to commit changes

aaronthebald committed:
Product is working as intended Using @binding you can now select an event. Update the details. and post it to firebase

aaronthebald commented:
Please let me know if there is anything you would like me to improve. Thank you so much for setting this project up!

mikaelacaron commented on Basic-Car-Maintenance.xcodeproj/project.pbxproj:
```diff
@@ -727,7 +731,7 @@
 				CODE_SIGN_STYLE = Automatic;
 				DEVELOPMENT_ASSET_PATHS = "\"Basic-Car-Maintenance/Preview Content\"";
-				DEVELOPMENT_TEAM = "";
+				DEVELOPMENT_TEAM = MW75NQ7TG5;
```
```suggestion
				DEVELOPMENT_TEAM ="";
```

mikaelacaron replied:
Click these "commit suggestion" buttons after you make your changes, it's someone with the project that I messed up and will be fixing today

mikaelacaron commented on Basic-Car-Maintenance.xcodeproj/project.pbxproj:
```diff
@@ -766,7 +770,7 @@
 				CODE_SIGN_STYLE = Automatic;
 				DEVELOPMENT_ASSET_PATHS = "\"Basic-Car-Maintenance/Preview Content\"";
-				DEVELOPMENT_TEAM = "";
+				DEVELOPMENT_TEAM = MW75NQ7TG5;
```
```suggestion
				DEVELOPMENT_TEAM = "";
```

mikaelacaron commented on Basic-Car-Maintenance.xcodeproj/project.pbxproj:
```diff
@@ -72,7 +73,7 @@
 		FFDADF8F2ACD35A200DDEF79 /* Embed Foundation Extensions */ = {
 			isa = PBXCopyFilesBuildPhase;
-			buildActionMask = 2147483647;
+			buildActionMask = 12;
```
Do you know what this is changing? If not, I'll just change it back

aaronthebald replied:
I have no idea. I'm not sure how it even was changed.

mikaelacaron commented on Basic-Car-Maintenance/Shared/Dashboard/ViewModels/DashboardViewModel.swift:
```diff
@@ -78,6 +78,21 @@ class DashboardViewModel: ObservableObject {
         }
     }
     
+    func updateEvent(_ maintenanceEvent: MaintenanceEvent) async {
+            
+            if let uid = authenticationViewModel.user?.uid {
+                guard let id = maintenanceEvent.id else {return}
+                var eventToUpdate = maintenanceEvent
+                eventToUpdate.userID = uid
+                try? Firestore
+                    .firestore()
+                    .collection("maintenance_events")
+                    .document(id)
+                    .setData(from: eventToUpdate)
+            }
+                    await self.getMaintenanceEvents()
+        }
+    
```
The indentation is off for inside this function, and can you handle the error that comes back from Firebase if the `setData` doesn't work? An alert should be shown

mikaelacaron commented on Basic-Car-Maintenance/Shared/Dashboard/Views/DashboardView.swift:
```diff
@@ -10,6 +10,10 @@ import SwiftUI
     
     @StateObject private var viewModel: DashboardViewModel
+    @State private var isShowingEditView = false
+        @State private var selectedMaintenanceEvent: MaintenanceEvent = MaintenanceEvent(
+            title: "", date: Date(), notes: ""
+        )
```
`selectedMaintenanceEvent` this should actually be optional rather than creating an object, because we're not always selecting an event

mikaelacaron commented on Basic-Car-Maintenance/Shared/Dashboard/Views/EditEventDetailView.swift:
```diff
@@ -0,0 +1,79 @@
+            })
+            .navigationTitle(Text("Add Maintenance"))
+            .toolbar {
+                ToolbarItem(placement: .topBarTrailing) {
```
These should be switched, the top trailing should be the update and the top leading should be the cancel

And put a single line break between the `ToolbarItem`s, so they're not right next to each other

mikaelacaron commented on Basic-Car-Maintenance/Shared/Dashboard/Views/EditEventDetailView.swift:
```diff
@@ -0,0 +1,79 @@
+            
+        }
+    }
+    func setMaintenanceEventValues(event: MaintenanceEvent) {
```
There should be a single line break before this function, so that it's not right next to the body

mikaelacaron commented on Basic-Car-Maintenance/Shared/Dashboard/Views/EditEventDetailView.swift:
```diff
@@ -0,0 +1,79 @@
+                }
+                ToolbarItem(placement: .topBarLeading) {
+                    Button {
+                        var event = MaintenanceEvent(title: title, date: date, notes: notes)
+                        event.id = selectedEvent.id
+                        print("button pressed")
+                        Task {
+                            await viewModel.updateEvent(event)
+                        }
+                        dismiss()
```
You don't need to create a new event object you should be able to pass `selectedEvent` directly, and given that you don't need 3 new state properties for `title`, `date`, and `notes`, you should be able to use `selectedEvent`

See below:
```swift
TextField("Title", text: $selectedEvent.title)
```

mikaelacaron replied:
And the `dismiss` should be inside of the `Task` so that the view doesn't dismiss if there was an error

aaronthebald replied:
First of all, thank you so much for the detailed review! I couldn't watch live but I just rewatched the stream.

As for this issue, when I attempted to update the selectedEvent values directly, I ran into errors because the MaintenanceEvent properties are let constants. So rather than refactoring the model, I created a new object with the updated values and then in firebase overwrote the old object with the new one with the updated values.

Is there another approach I should use instead? 

I appreciate all your feedback.

mikaelacaron replied:
I'd say go ahead and change the model 👍

mikaelacaron commented on Basic-Car-Maintenance/Shared/Dashboard/Views/EditEventDetailView.swift:
```diff
@@ -0,0 +1,79 @@
+struct EditMaintenanceEventView: View {
+    @Binding var selectedEvent: MaintenanceEvent
+    @ObservedObject var viewModel: DashboardViewModel
+    @State private var title = ""
+    @State private var date = Date()
+    @State private var notes = ""
```
We don't need these, see other comment below

mikaelacaron commented on Basic-Car-Maintenance/Shared/Dashboard/Views/EditEventDetailView.swift:
```diff
@@ -0,0 +1,79 @@
+        NavigationStack {
+            Form {
+                Section {
+                    TextField("Title",
+                              text: $title)
```
This should be on one line

mikaelacaron commented on Basic-Car-Maintenance/Shared/Dashboard/Views/EditEventDetailView.swift:
```diff
@@ -0,0 +1,79 @@
+//
+//  EditEventDetailView.swift
```
This comment should match the view's name `EditMaintenanceEventView`

mikaelacaron reviewed (changes requested):
Great work!! I have a few changes about indentation and line breaks, also a different way to do the editing on the `EditMaintenanceEventView`

Please make these changes and click "Resolve" on the ones you fix, or if you have questions about those, and once you're done, click the re-review button

<img width="345" alt="Re-review button" src="https://github.com/mikaelacaron/Basic-Car-Maintenance/assets/22946902/9a227eca-4ebe-4f7e-bde5-2acd478eca97">

aaronthebald committed:
Handled error if updateMaintenanceEvent() fails.
Fixed formatting issues in several places

aaronthebald committed:
updated the var selectedMaintenanceEvent to an optional var in both the Dashboard view as well as the EditMaintenanceEventView.
added guard statements to safely unwrap the value when attempting to access it.

aaronthebald committed:
swapped placement of the ToolBarItems on the EditMaintenanceEventView.

mikaelacaron committed:
Merge branch 'dev' into 4-feature-edit-maintenance-event-detail

mikaelacaron commented on Basic-Car-Maintenance/Shared/Dashboard/ViewModels/DashboardViewModel.swift:
```diff
@@ -78,6 +78,26 @@ class DashboardViewModel: ObservableObject {
+                errorMessage = error.localizedDescription
+            }
+        }
+        await self.getMaintenanceEvents()
```
This could actually be moved into the do catch, because if there was an error, there's no need to fetch new data, because nothing would be new

mikaelacaron commented on Basic-Car-Maintenance/Shared/Dashboard/Views/DashboardView.swift:
```diff
@@ -36,6 +38,20 @@ struct DashboardView: View {
+                            selectedMaintenanceEvent = event
+                            isShowingEditView = true
+                        } label: {
+                            VStack {
+                                Text("Edit")
+                                Image(systemName: "pencil")
+                            }
```
You can achieve the same thing with a `Label`, and it's better accessibility wise

`Label("Edit", systemImage: "pencil")`

mikaelacaron commented on Basic-Car-Maintenance/Shared/Dashboard/Views/DashboardView.swift:
```diff
@@ -36,6 +38,20 @@ struct DashboardView: View {
+                            }
+                        }
+                    }
+                    .sheet(isPresented: $isShowingEditView) {
```
Just a note the sheet can be moved on the same level of modifiers as the main stuff (it doesn't have to specifically be here

mikaelacaron commented on Basic-Car-Maintenance.xcodeproj/project.pbxproj:
```diff
@@ -71,7 +72,7 @@
 		FFDADF8F2ACD35A200DDEF79 /* Embed Foundation Extensions */ = {
 			isa = PBXCopyFilesBuildPhase;
-			buildActionMask = 2147483647;
+			buildActionMask = 12;
```
```suggestion
			buildActionMask = 2147483647;
```

mikaelacaron reviewed (approved):
Great work!! I'm going to go ahead and merge this, all the other comments I made are just side notes

mikaelacaron committed:
Update project.pbxproj

aaronthebald commented:
Thank you! Looking forward to the rest of the month
